test/test_piece_picker.cpp

```diff
@@ -885,64 +885,72 @@ TORRENT_TEST(cursors_sweep_in_we_have)
 	TEST_CHECK(!p->is_seeding());
 	TEST_EQUAL(p->cursor(), piece_index_t(0));
 	TEST_EQUAL(p->reverse_cursor(), piece_index_t(7));
+	p->set_sequential_range(piece_index_t(1), piece_index_t(5));
+	for (piece_index_t i(1); i < piece_index_t(6); ++i)
+	{
+		TEST_EQUAL(p->cursor(), piece_index_t(i));
+		TEST_EQUAL(p->reverse_cursor(), piece_index_t(6));
+		p->we_have(i);
+	}
+	TEST_EQUAL(p->cursor(), piece_index_t(7));
+	TEST_EQUAL(p->reverse_cursor(), piece_index_t(0));
+	p->we_have(piece_index_t(0));
+	p->we_have(piece_index_t(6));
+	TEST_CHECK(p->is_finished());
+	TEST_CHECK(p->is_seeding());
 }
 
-TORRENT_TEST(cursors_sweep_down_we_dont_have)
+TORRENT_TEST(cursors_sweep_down_seq_range_we_have)
 {
-	auto p = setup_picker("7654321", "*******", "", "");
-	TEST_CHECK(p->is_finished());
-	TEST_CHECK(p->is_seeding());
-	TEST_EQUAL(p->cursor(), piece_index_t(7));
-	TEST_EQUAL(p->reverse_cursor(), piece_index_t(0));
-	for (piece_index_t i(6); i >= piece_index_t(0); --i)
-	{
-		p->we_dont_have(i);
-		TEST_EQUAL(p->cursor(), i);
-		TEST_EQUAL(p->reverse_cursor(), piece_index_t(7));
-	}
+	auto p = setup_picker("7654321", "       ", "", "");
 	TEST_CHECK(!p->is_finished());
 	TEST_CHECK(!p->is_seeding());
 	TEST_EQUAL(p->cursor(), piece_index_t(0));
 	TEST_EQUAL(p->reverse_cursor(), piece_index_t(7));
+	p->set_sequential_range(piece_index_t(1), piece_index_t(5));
+	for (piece_index_t i(5); i >= piece_index_t(1); --i)
+	{
+		TEST_EQUAL(p->cursor(), piece_index_t(1));
+		TEST_EQUAL(p->reverse_cursor(), next(i));
+		p->we_have(i);
+	}
+	TEST_EQUAL(p->cursor(), piece_index_t(7));
+	TEST_EQUAL(p->reverse_cursor(), piece_index_t(0));
+	p->we_have(piece_index_t(0));
+	p->we_have(piece_index_t(6));
+	TEST_CHECK(p->is_finished());
+	TEST_CHECK(p->is_seeding());
 }
 
-TORRENT_TEST(cursors_sweep_out_we_dont_have)
+TORRENT_TEST(cursors_sweep_in_seq_range_we_have)
 {
-	auto p = setup_picker("7654321", "*******", "", "");
-	TEST_CHECK(p->is_finished());
-	TEST_CHECK(p->is_seeding());
-	TEST_EQUAL(p->cursor(), piece_index_t(7));
-	TEST_EQUAL(p->reverse_cursor(), piece_index_t(0));
-	for (piece_index_t left(3), right(3);
-		left >= piece_index_t(0) && right < piece_index_t(7);
-		--left, ++right)
-	{
-		p->we_dont_have(left);
-		p->we_dont_have(right);
-		TEST_EQUAL(p->cursor(), left);
-		TEST_EQUAL(p->reverse_cursor(), next(right));
-	}
+	auto p = setup_picker("7654321", "       ", "", "");
 	TEST_CHECK(!p->is_finished());
 	TEST_CHECK(!p->is_seeding());
 	TEST_EQUAL(p->cursor(), piece_index_t(0));
 	TEST_EQUAL(p->reverse_cursor(), piece_index_t(7));
+	p->set_sequential_range(piece_index_t(1), piece_index_t(5));
+	for (piece_index_t left(1), right(5); left <= piece_index_t(3)
+		&& right >= piece_index_t(3); ++left, --right)
+	{
+		TEST_EQUAL(p->cursor(), left);
+		TEST_EQUAL(p->reverse_cursor(), next(right));
+		p->we_have(left);
+		p->we_have(right);
+	}
+	TEST_EQUAL(p->cursor(), piece_index_t(7));
+	TEST_EQUAL(p->reverse_cursor(), piece_index_t(0));
```

Comment:
Perhaps I'm missing something. I don't understand why it makes sense to set `cursor` to 7 and `reverse_cursor` to 0. At this point the range isn't valid anymore, or being used. But start > end is an invalid range. This is currently a special case to indicate we're a seed, but in this case we're not a seed.

It would seem to make more sense to set them back to [0, 7] when they meet in the middle.

Reply:
Currently with the sequential download functionality, when the cursors cross that means sequential downloading is finished and so they are set to cursor at the end and reverse cursor at the beginning as their end state. The code throughout the piece picker does this and numerous conditions assume this end state. I'm just following that convention.

Reply:
Also if you were to set them back to the beginning and end of the torrent then they will be starting to sweep again, as if they haven't downloaded anything. As though their state machine would never finish and loop forever. What end state would there be to stop them? Another flag? Or automatically setting sequential download mode to false? Or something else?

That is why I left the special case end state as it is. Yes it is an invalid range and doesn't make sense in and of itself, but it is a way to indicate the end state without introducing more state, which is something I recall you are not fond of.

Reply:
ok, so `reverse_cursor = 0` and `m_cursor = num_pieces` is a special case to mean sequential downloading is disabled. IS that right? When in this state, the piece picker will ignore the `sequential` flag, is that right?

I think this should be documented in comments. I don't think this behavior is obvious (but I think it's the right one).

Reply:
there should also be a test demonstrating this

Reply:
Yes when they are in that state that means sequential download has ended. Though how would I check this? Through an invariant assertion or through a subtest?